Complete validate Transaction with

lomamech opened the pull request:
Complete validate Transaction with  …
- check single line transaction
- check mising accounts
- check missing dates
- check Date in wrong period
- check different date

closes #1498

lomamech committed:
Complete validate Transaction with
- check single line transaction
- check mising accounts
- check missing dates
- check Date in wrong period
- check different date

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -303,24 +303,81 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+
+
+        // Check if they are account number Null
+        if(!row.account_number){
```
A lot of these checks you can simplify by removing the "if" and setting the error equal to the check.  LIke this:
```js
accountNull = !row.account_number;
dateNull = !row.trans_date;
dateDifference = Dates.util.str(row.trans_date) !== Dates.util.str(initialDate);
```
and so on.  This  might reduce your code a bit.

The comments are nice!

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -303,24 +303,81 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+      });
+
+      if(accountNull) {
+        error = 'POSTING_JOURNAL.ERRORS.MISSING_ACCOUNTS';
```
Instead of having these long text strings here, it would be nicer to list them all at the top of the module, like is done with `ROW_EDIT_FLAG` and others at the top of this file (https://github.com/IMA-WorldHealth/bhima-2.X/pull/1501/files#diff-2b8e7d8131d2079c84c7b41baa2361a5R25).  Something like this:
```js
var ERR_MISSING_ACCOUNTS = 'POSTING_JOURNAL.ERRORS.MISSING_ACCOUNTS';
var ERR_MISSING_DATES = 'POSTING_JOURNAL.ERRORS.MISSING_DATES';
//....

// much later...

if (accountNull) {
 error = ERR_MISSING_ACCOUNTS;
} else if { /// etc...
```

This will make it much easier to update translation keys in the future - ideally these keys will be shared with the complex vouchers module.

jniles commented on client/src/modules/journal/journal.js:
```diff
@@ -313,11 +313,12 @@ function JournalController(Journal, Sorting, Grouping, Filtering, Columns, Confi
-        if(error === 'POSTING_JOURNAL.ERRORS.UNBALANCED_TRANSACTIONS'){
-          Notify.warn(error);
-        } else {
+        if(error.status){
```
You've mixed `if(` and `if (` in the PR a few times.  ESLint prefers that you use `if (` to distinguish it from a function call.  Could you use `if (` instead of `if(`?

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -303,24 +303,81 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+      }
+        
+
+      transaction.forEach(function (row) {
```
One loop is better than many! :+1:

jniles commented on client/src/js/services/grid/TransactionService.js:
```diff
@@ -303,24 +303,81 @@ function TransactionService($timeout, util, uiGridConstants, bhConstants, Notify
+    // If the transaction are single line transaction 
+    if(numberOfLine === 1){
+      error = 'POSTING_JOURNAL.ERRORS.SINGLE_LINE_TRANSACTION';
+    } else {
```
Instead of doing this giant `if () else {}`, you can just return here - if there is only one line, it is clearly a single line transaction and we don't need to do anything else.  It will save you one indentation.

Like this:
```js
// as suggested earlier...
var ERR_SINGLE_LINE_TRANSACTION = 'POSTING_JOURNAL.ERRORS.SINGLE_LINE_TRANSACTION';

// and then here....

if (numberOfLines === 1) {
  return ERR_SINGLE_LINE_TRANSACTION;
}

var debit = 0;
var credit = 0;
/// ... etc...
```

jniles reviewed (commented):
@lomamech looks pretty good!  Good work.

I've made a few suggestions to keep this code maintainable and readable.  The logic seems correct to me. 👍

lomamech committed:
- Improve code for Validate Transaction
- Adding validation for prevent debit and credit null
- Adding validation for prevent debited and credited on the same line
